qutebrowser/completion/completer.py

```diff
@@ -255,136 +257,34 @@ def _update_completion(self):
-            if cursor_pos <= len(part):
-                # foo| bar
-                self._cursor_part = i - skip
-                if spaces:
-                    self._empty_item_idx = i - skip
-                else:
-                    self._empty_item_idx = None
-                log.completion.vdebug("cursor_pos {} <= len(part) {}, "
-                                      "setting cursor_part {} - {} (skip), "
-                                      "empty_item_idx {}".format(
-                                          cursor_pos, len(part), i, skip,
-                                          self._empty_item_idx))
-                break
-            cursor_pos -= len(part)
-            log.completion.vdebug(
-                "Removing len({!r}) -> {} from cursor_pos -> {}".format(
-                    part, len(part), cursor_pos))
-        else:
-            if i == 0:
-                # Initial `:` press without any text.
-                self._cursor_part = 0
-            else:
-                self._cursor_part = i - skip
-            if spaces:
-                self._empty_item_idx = i - skip
-            else:
-                self._empty_item_idx = None
-        log.completion.debug("cursor_part {}, spaces {}".format(
-            self._cursor_part, spaces))
-        return
-
-    def _change_completed_part(self, newtext, immediate=False):
+    def _change_completed_part(self, newtext, before, after, immediate=False):
         """Change the part we're currently completing in the commandline.
 
         Args:
-            text: The text to set (string).
+            text: The text to set (string) for the token under the cursor.
+            before: Commandline tokens before the token under the cursor.
+            after: Commandline tokens after the token under the cursor.
             immediate: True if the text should be completed immediately
                        including a trailing space and we shouldn't continue
                        completing the current item.
         """
-        parts = self._split()
-        log.completion.debug("changing part {} to '{}'".format(
-            self._cursor_part, newtext))
-        try:
-            parts[self._cursor_part] = newtext
-        except IndexError:
-            parts.append(newtext)
-        # We want to place the cursor directly after the part we just changed.
-        cursor_str = self._cmd.prefix() + ' '.join(
-            parts[:self._cursor_part + 1])
-        if immediate:
-            # If we should complete immediately, we want to move the cursor by
-            # one more char, to get to the next field.
-            cursor_str += ' '
-        text = self._cmd.prefix() + ' '.join(parts)
-        if immediate and self._cursor_part == len(parts) - 1:
-            # If we should complete immediately and we're completing the last
-            # part in the commandline, we automatically add a space.
+        text = self._cmd.prefix() + ' '.join([*before, newtext])
```

Comment:
This will only work in Python 3.5, but you can probably do `before + [newtext]` instead?